### PR TITLE
[RFC][dynamo] add decorator to register polyfill for unsupported C++ function to avoid graph break

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133895
* #133894
* #133876
* #133875
* #133864
* #133977
* #133862
* #133822
* #133801
* #133771
* #133779
* #133778
* #133769
* __->__ #133712

Add decorator `torch.compiler.substitute_in_graph` to register polyfill for unsupported C++ function to avoid graph break. This API provides an official way to add support for dynamo for third-party C extensions. Also, it can be used to simplify our implementation for `torch._dynamo.polyfill`.

https://github.com/pytorch/pytorch/blob/5ee070266fb39916559ae773272c57521397eb1a/torch/_dynamo/variables/builtin.py#L97-L107

Example:

```python
>>> import operator
>>> operator.indexOf([1, 2, 3, 4, 5], 3)
2

>>> torch.compile(operator.indexOf, fullgraph=True)([1, 2, 3, 4, 5], 3)
Unsupported: ...

>>> @torch.compiler.substitute_in_graph(operator.indexOf)
... def indexOf(sequence, x):
...     for i, item in enumerate(sequence):
...         if item is x or item == x:
...             return i
...     raise ValueError("sequence.index(x): x not in sequence")

>>> torch.compile(operator.indexOf, fullgraph=True)([1, 2, 3, 4, 5], 3)
2
```

cc @rec @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @xmfan

### CODE REVIEW

File: torch/compiler/__init__.py [hunk 1]
```diff
@@ -231,6 +284,7 @@ def is_compiling() -> bool:
     else:
         return _is_compiling_flag
 
+
```
Comment (ezyang) [inline torch/compiler/__init__.py:287]: What's with the white space changes

Comment (XuehaiPan) [reply torch/compiler/__init__.py:287]: I removed `torch/compiler/__init__.py` from the `PYFMT` exclude pattern.

**Review by jansel — CHANGES_REQUESTED** [review]
I like this idea.

Can we change the existing polyfill functions to use this API?

**Comment by XuehaiPan** [comment]
> Can we change the existing polyfill functions to use this API?

Yes, we can.

**Comment by XuehaiPan** [comment]
I added one example in the test in this PR and two examples in follow-up PRs to demonstrate how the new API can be used.

File: torch/_dynamo/decorators.py [hunk 2]
```diff
@@ -159,6 +164,106 @@ def forbid_in_graph(fn):
...
+
+    def wrapper(traceable_fn: _F) -> _F:
+        from torch._dynamo.guards import GuardBuilder
+        from torch._dynamo.trace_rules import get_torch_obj_rule_map
+        from torch._dynamo.variables import PolyfilledFunctionVariable
+        from torch._dynamo.variables.builder import VariableBuilder
+
+        id_dispatch_map = VariableBuilder._id_dispatch()
+        if id(original_fn) in id_dispatch_map:
+            raise RuntimeError(
```
Comment (Skylion007) [inline torch/_dynamo/decorators.py:223]: Nit should be value error right?

**Review by Skylion007 — DISMISSED** [review]

File: torch/_dynamo/decorators.py (hunk 2 above)
Comment (XuehaiPan) [reply torch/_dynamo/decorators.py]: updated

**Review by jansel — APPROVED** [review]

**Comment by jeanschmidt** [comment]
Duplicating it here:
This stack seems to have introduced:

```
------------------------------ Captured log call ------------------------------
ERROR    numba.cuda.cudadrv.driver:driver.py:392 Call to cuInit results in CUDA_ERROR_UNKNOWN
__________ TestUnflatten.test_unflattened_module_nodes_has_meta_val ___________
```

for windows tests `[win-vs2019-cpu-py3 / test (default, 3, 3, windows.4xlarge.nonephemeral)](https://github.com/pytorch/pytorch/actions/runs/10461735144/job/28977363201#logs)`

**Comment by ZainRizvi** [comment]
@pytorchbot revert -m "breaking main windows cpu tests - this stack still causes that windows test to fail" -c weird

Since the stack was based off of a commit from `trunk` that already had that test failing on it (from before the last revert), mergebot incorrectly thought that the test was a flaky failure (it's marked as such in https://github.com/pytorch/pytorch/pull/133771)

**Comment by XuehaiPan** [comment]
Related:

- #133987

**Pull request closed without merging**